File: src/services/authEvents.ts
import { invoke } from '@tauri-apps/api/core'

export interface AuthEvent {
  type: 'SessionUnlocked' | 'SessionLocked' | 'AutoUnlockAttempted'
  timestamp: string
  // SessionUnlocked
  via_auto_unlock?: boolean
  // SessionLocked
  reason?: {
    Manual?: null
    Timeout?: null
    Error?: string
  }
  // AutoUnlockAttempted
  success?: boolean
  error?: string | null
}

export interface AuthSession {
  isSetup: boolean
  isUnlocked: boolean
  autoUnlockEnabled: boolean
  keychainAvailable: boolean
  sessionActive: boolean
  sessionExpiresAt?: string
  loadedDeviceCount: number
  timestamp: string
}

export interface AuthEventsResponse {
  events: AuthEvent[]
  timestamp: string
}

/**
 * Service for handling auth events and session management
 */
class AuthEventsService {
  private pollingInterval: number | null = null
  private statusListeners: Set<(status: AuthSession) => void> = new Set()

  /**
   * Start polling for auth events
   */
  async startPolling(intervalMs: number = 2000): Promise<void> {
    if (this.pollingInterval) {
      this.stopPolling()
    }

    this.pollingInterval = window.setInterval(async () => {
      try {
        // Get latest auth session status
        const status = await this.getAuthSessionStatus()

        // Notify status listeners
        this.statusListeners.forEach(listener => listener(status))

        // Note: For now we're not implementing event polling since we don't store events in backend
        // The status polling above is sufficient for most use cases
      } catch (error) {
        console.error('Failed to poll auth status:', error)
      }
    }, intervalMs)
  }

  /**
   * Stop polling for auth events
   */
  stopPolling(): void {
    if (this.pollingInterval) {
      clearInterval(this.pollingInterval)
      this.pollingInterval = null
    }
  }

  /**
   * Get current auth session status
   */
  async getAuthSessionStatus(): Promise<AuthSession> {
    return await invoke<AuthSession>('get_auth_session_status')
  }

  /**
   * Notify backend that session was unlocked
   */
  async notifySessionUnlocked(): Promise<void> {
    await invoke('notify_session_unlocked')
  }

  /**
   * Notify backend that session was locked
   */
  async notifySessionLocked(reason: string = 'manual'): Promise<void> {
    await invoke('notify_session_locked', { reason })
  }

  /**
   * Subscribe to auth events (currently uses polling)
   */
  async subscribeAuthEvents(): Promise<void> {
    await invoke('subscribe_auth_events')
  }

  /**
   * Get auth events since timestamp
   */
  async getAuthEvents(since?: string): Promise<AuthEventsResponse> {
    return await invoke<AuthEventsResponse>('get_auth_events', { since })
  }



  /**
   * Add status change listener
   */
  addStatusListener(listener: (status: AuthSession) => void): void {
    this.statusListeners.add(listener)
  }

  /**
   * Remove status change listener
   */
  removeStatusListener(listener: (status: AuthSession) => void): void {
    this.statusListeners.delete(listener)
  }

  /**
   * Cleanup - stop polling and clear listeners
   */
  cleanup(): void {
    this.stopPolling()
    this.statusListeners.clear()
  }
}

// Export singleton instance
export const authEventsService = new AuthEventsService()

File: src-tauri/src/commands/auth_events.rs
use crate::core::auth_session_manager::SessionLockReason;
use crate::state::AppState;
use chrono::Utc;
use serde_json::json;
use tauri::State;

/// Manually trigger auth session unlock notification (called after successful unlock)
#[tauri::command]
pub async fn notify_session_unlocked(state: State<'_, AppState>) -> Result<(), String> {
    let auth_session_manager = state.auth_session_manager.lock().await;
    auth_session_manager
        .on_session_unlocked()
        .await
        .map_err(|e| e.to_string())?;

    Ok(())
}

/// Manually trigger auth session lock notification (called after manual lock)
#[tauri::command]
pub async fn notify_session_locked(
    state: State<'_, AppState>,
    reason: Option<String>,
) -> Result<(), String> {
    let lock_reason = match reason.as_deref() {
        Some("manual") => SessionLockReason::Manual,
        Some("timeout") => SessionLockReason::Timeout,
        Some(other) => SessionLockReason::Error(other.to_string()),
        None => SessionLockReason::Manual,
    };

    let auth_session_manager = state.auth_session_manager.lock().await;
    auth_session_manager
        .on_session_locked(lock_reason)
        .await
        .map_err(|e| e.to_string())?;

    Ok(())
}

/// Get current auth session status
#[tauri::command]
pub async fn get_auth_session_status(
    state: State<'_, AppState>,
) -> Result<serde_json::Value, String> {
    let database_service = state.database_service.lock().await;
    let status = database_service
        .get_master_password_status()
        .await
        .map_err(|e| e.to_string())?;

    Ok(json!({
        "isSetup": status.is_setup,
        "isUnlocked": status.is_unlocked,
        "autoUnlockEnabled": status.auto_unlock_enabled,
        "keychainAvailable": status.keychain_available,
        "sessionActive": status.session_active,
        "sessionExpiresAt": status.session_expires_at,
        "loadedDeviceCount": status.loaded_device_count,
        "timestamp": Utc::now()
    }))
}
